---
title: 失敗時に補償アクションを伴う操作のシーケンス
excerpt: Effectは、前のステップに依存する連続操作を促進し、失敗時にはロールバックを行います。
bottomNavigation: pagination
---

特定のシナリオでは、各操作の成功が前の操作に依存する一連の連鎖操作を実行する必要があるかもしれません。しかし、いずれかの操作が失敗した場合、以前のすべての成功した操作の効果を逆転させたいと考えるでしょう。このパターンは、すべての操作が成功するか、または何も効果がないことを保証する必要がある場合に価値があります。

Effectは、[Effect.acquireRelease](../scope#defining-resources)関数と[Exit](../../../other/data-types/exit)型を組み合わせてこのパターンを実現する方法を提供します。
[Effect.acquireRelease](../scope#defining-resources)関数は、リソースを取得し、それを使用して操作を行い、完了したらリソースを解放することを可能にします。
[Exit](../../../other/data-types/exit)型は、効果的な計算の結果を表し、それが成功したか失敗したかを示します。

このパターンを実装する例を見てみましょう。アプリケーションで「ワークスペース」を作成する場合を考えます。これは、S3バケット、ElasticSearchインデックス、および前の2つに依存するデータベースエントリを作成することを含みます。

まず、必要な[サービス](./context-management/services)のドメインモデルを定義します：`S3`、`ElasticSearch`、および`Database`。

```twoslash include Services
import { Effect, Context } from "effect"

export class S3Error {
  readonly _tag = "S3Error"
}

export interface Bucket {
  readonly name: string
}

export class S3 extends Context.Tag("S3")<
  S3,
  {
    readonly createBucket: Effect.Effect<Bucket, S3Error>
    readonly deleteBucket: (bucket: Bucket) => Effect.Effect<void>
  }
>() {}

export class ElasticSearchError {
  readonly _tag = "ElasticSearchError"
}

export interface Index {
  readonly id: string
}

export class ElasticSearch extends Context.Tag("ElasticSearch")<
  ElasticSearch,
  {
    readonly createIndex: Effect.Effect<Index, ElasticSearchError>
    readonly deleteIndex: (index: Index) => Effect.Effect<void>
  }
>() {}

export class DatabaseError {
  readonly _tag = "DatabaseError"
}

export interface Entry {
  readonly id: string
}

export class Database extends Context.Tag("Database")<
  Database,
  {
    readonly createEntry: (
      bucket: Bucket,
      index: Index
    ) => Effect.Effect<Entry, DatabaseError>
    readonly deleteEntry: (entry: Entry) => Effect.Effect<void>
  }
>() {}
```

```ts filename="Services.ts" twoslash
// @include: Services
```

次に、Workspaceのための3つの作成アクションと全体のトランザクション（`make`）を定義します。

<Tabs items={["Effect.genを使用", "pipeを使用"]}>
<Tab>

```ts filename="Workspace.ts" twoslash
// @filename: Services.ts
// @include: Services

// @filename: Workspace.ts
// ---cut---
import { Effect, Exit } from "effect"
import * as Services from "./Services"

// バケットを作成し、操作が失敗した場合にバケットを削除するリリース関数を定義します。
const createBucket = Effect.gen(function* () {
  const { createBucket, deleteBucket } = yield* Services.S3
  return yield* Effect.acquireRelease(createBucket, (bucket, exit) =>
    // Effect.acquireRelease操作のリリース関数は、メインのエフェクトが完了した後に取得したリソース（バケット）を処理する責任があります。
    // メインのエフェクトが成功したか失敗したかに関わらず呼び出されます。
    // メインのエフェクトが失敗した場合、Exit.isFailure(exit)はtrueとなり、関数はdeleteBucket(bucket)を呼び出してロールバックを行います。
    // メインのエフェクトが成功した場合、Exit.isFailure(exit)はfalseとなり、関数はEffect.voidを返します。これは成功したが何もしないエフェクトを表します。
    Exit.isFailure(exit) ? deleteBucket(bucket) : Effect.void
  )
})

// インデックスを作成し、操作が失敗した場合にインデックスを削除するリリース関数を定義します。
const createIndex = Effect.gen(function* () {
  const { createIndex, deleteIndex } = yield* Services.ElasticSearch
  return yield* Effect.acquireRelease(createIndex, (index, exit) =>
    Exit.isFailure(exit) ? deleteIndex(index) : Effect.void
  )
})

// データベースにエントリを作成し、操作が失敗した場合にエントリを削除するリリース関数を定義します。
const createEntry = (bucket: Services.Bucket, index: Services.Index) =>
  Effect.gen(function* () {
    const { createEntry, deleteEntry } = yield* Services.Database
    return yield* Effect.acquireRelease(
      createEntry(bucket, index),
      (entry, exit) =>
        Exit.isFailure(exit) ? deleteEntry(entry) : Effect.void
    )
  })

export const make = Effect.scoped(
  Effect.gen(function* () {
    const bucket = yield* createBucket
    const index = yield* createIndex
    return yield* createEntry(bucket, index)
  })
)
```

</Tab>
<Tab>

```ts filename="Workspace.ts" twoslash
// @filename: Services.ts
// @include: Services

// @filename: Workspace.ts
// ---cut---
import { Effect, Exit } from "effect"
import * as Services from "./Services"

// バケットを作成し、操作が失敗した場合にバケットを削除するリリース関数を定義します。
const createBucket = Services.S3.pipe(
  Effect.andThen(({ createBucket, deleteBucket }) =>
    Effect.acquireRelease(createBucket, (bucket, exit) =>
      // Effect.acquireRelease操作のリリース関数は、メインのエフェクトが完了した後に取得したリソース（バケット）を処理する役割を担います。
      // メインのエフェクトが成功したか失敗したかに関わらず、この関数は呼び出されます。
      // メインのエフェクトが失敗した場合、Exit.isFailure(exit)はtrueとなり、関数はdeleteBucket(bucket)を呼び出してロールバックを行います。
      // メインのエフェクトが成功した場合、Exit.isFailure(exit)はfalseとなり、関数はEffect.voidを返します。これは成功したが何もしないエフェクトを表します。
      Exit.isFailure(exit) ? deleteBucket(bucket) : Effect.void
    )
  )
)

// インデックスを作成し、操作が失敗した場合にインデックスを削除するリリース関数を定義します。
const createIndex = Services.ElasticSearch.pipe(
  Effect.andThen(({ createIndex, deleteIndex }) =>
    Effect.acquireRelease(createIndex, (index, exit) =>
      Exit.isFailure(exit) ? deleteIndex(index) : Effect.void
    )
  )
)

// データベースにエントリを作成し、操作が失敗した場合にエントリを削除するリリース関数を定義します。
const createEntry = (bucket: Services.Bucket, index: Services.Index) =>
  Services.Database.pipe(
    Effect.andThen(({ createEntry, deleteEntry }) =>
      Effect.acquireRelease(createEntry(bucket, index), (entry, exit) =>
        Exit.isFailure(exit) ? deleteEntry(entry) : Effect.void
      )
    )
  )

export const make = Effect.scoped(
  Effect.Do.pipe(
    Effect.bind("bucket", () => createBucket),
    Effect.bind("index", () => createIndex),
    Effect.andThen(({ bucket, index }) => createEntry(bucket, index))
  )
)
```

</Tab>
</Tabs>

ワークスペースコードの動作をテストするために、シンプルなサービス実装を作成します。
これを実現するために、[レイヤー](./context-management/layers)を利用してテストサービスを構築します。
これらのレイヤーは、エラーを含むさまざまなシナリオに対応でき、`FailureCase`タイプを使用して制御することができます。

<Tabs items={["Effect.genを使用する", "pipeを使用する"]}>
<Tab>

```ts filename="WorkspaceTest.ts" twoslash
// @filename: Services.ts
// @include: Services

// @filename: Workspace.ts
import { Effect, Exit } from "effect"
import * as Services from "./Services"

const createBucket = Services.S3.pipe(
  Effect.andThen(({ createBucket, deleteBucket }) =>
    Effect.acquireRelease(createBucket, (bucket, exit) =>
      Exit.isFailure(exit) ? deleteBucket(bucket) : Effect.void
    )
  )
)

const createIndex = Services.ElasticSearch.pipe(
  Effect.andThen(({ createIndex, deleteIndex }) =>
    Effect.acquireRelease(createIndex, (index, exit) =>
      Exit.isFailure(exit) ? deleteIndex(index) : Effect.void
    )
  )
)

const createEntry = (bucket: Services.Bucket, index: Services.Index) =>
  Services.Database.pipe(
    Effect.andThen(({ createEntry, deleteEntry }) =>
      Effect.acquireRelease(createEntry(bucket, index), (entry, exit) =>
        Exit.isFailure(exit) ? deleteEntry(entry) : Effect.void
      )
    )
  )

export const make = Effect.scoped(
  Effect.Do.pipe(
    Effect.bind("bucket", () => createBucket),
    Effect.bind("index", () => createIndex),
    Effect.andThen(({ bucket, index }) => createEntry(bucket, index))
  )
)

// @filename: WorkspaceTest.ts
// ---cut---
import { Effect, Context, Layer, Console } from "effect"
import * as Services from "./Services"
import * as Workspace from "./Workspace"

// `FailureCaseLiterals`型を使用すると、サービスをテストする際にさまざまなエラーシナリオを提供できます。
// 例えば、値として「S3」を指定すると、S3サービスに特有のエラーシナリオをシミュレートできます。
// これにより、プログラムがエラーを正しく処理し、さまざまな状況で期待通りに動作することを確認できます。
// 同様に、「ElasticSearch」や「Database」などの値を指定して、それらのサービスのエラーシナリオをシミュレートすることもできます。
// エラーが発生しない場合をテストしたい場合は、`undefined`を指定できます。
// このパラメータを使用することで、サービスを徹底的にテストし、さまざまなエラー条件下での動作を検証できます。
type FailureCaseLiterals = "S3" | "ElasticSearch" | "Database" | undefined

class FailureCase extends Context.Tag("FailureCase")<
  FailureCase,
  FailureCaseLiterals
>() {}

// S3サービスのテストレイヤーを作成する

const S3Test = Layer.effect(
  Services.S3,
  Effect.gen(function* () {
    const failureCase = yield* FailureCase
    return {
      createBucket: Effect.gen(function* () {
        console.log("[S3] creating bucket")
        if (failureCase === "S3") {
          return yield* Effect.fail(new Services.S3Error())
        } else {
          return { name: "<bucket.name>" }
        }
      }),
      deleteBucket: (bucket) =>
        Console.log(`[S3] delete bucket ${bucket.name}`)
    }
  })
)

// ElasticSearchサービスのテストレイヤーを作成する

const ElasticSearchTest = Layer.effect(
  Services.ElasticSearch,
  Effect.gen(function* () {
    const failureCase = yield* FailureCase
    return {
      createIndex: Effect.gen(function* () {
        console.log("[ElasticSearch] creating index")
        if (failureCase === "ElasticSearch") {
          return yield* Effect.fail(new Services.ElasticSearchError())
        } else {
          return { id: "<index.id>" }
        }
      }),
      deleteIndex: (index) =>
        Console.log(`[ElasticSearch] delete index ${index.id}`)
    }
  })
)

// Databaseサービスのテストレイヤーを作成する

const DatabaseTest = Layer.effect(
  Services.Database,
  Effect.gen(function* () {
    const failureCase = yield* FailureCase
    return {
      createEntry: (bucket, index) =>
        Effect.gen(function* () {
          console.log(
            `[Database] creating entry for bucket ${bucket.name} and index ${index.id}`
          )
          if (failureCase === "Database") {
            return yield* Effect.fail(new Services.DatabaseError())
          } else {
            return { id: "<entry.id>" }
          }
        }),
      deleteEntry: (entry) =>
        Console.log(`[Database] delete entry ${entry.id}`)
    }
  })
)

// S3、ElasticSearch、およびDatabaseサービスのすべてのテストレイヤーを単一のレイヤーにマージする

const layer = Layer.mergeAll(S3Test, ElasticSearchTest, DatabaseTest)

// ワークスペースコードをテストするための実行可能なエフェクトを作成する
// エフェクトにはテストレイヤーと未定義の値（失敗ケースなし）のFailureCaseサービスが提供される

const runnable = Workspace.make.pipe(
  Effect.provide(layer),
  Effect.provideService(FailureCase, undefined)
)

Effect.runPromise(Effect.either(runnable)).then(console.log)
```

</Tab>
<Tab>

```ts filename="Workspace.ts" twoslash
// @filename: Services.ts
// @include: Services

// @filename: Workspace.ts
import { Effect, Exit } from "effect"
import * as Services from "./Services"

const createBucket = Services.S3.pipe(
  Effect.andThen(({ createBucket, deleteBucket }) =>
    Effect.acquireRelease(createBucket, (bucket, exit) =>
      Exit.isFailure(exit) ? deleteBucket(bucket) : Effect.void
    )
  )
)

const createIndex = Services.ElasticSearch.pipe(
  Effect.andThen(({ createIndex, deleteIndex }) =>
    Effect.acquireRelease(createIndex, (index, exit) =>
      Exit.isFailure(exit) ? deleteIndex(index) : Effect.void
    )
  )
)

const createEntry = (bucket: Services.Bucket, index: Services.Index) =>
  Services.Database.pipe(
    Effect.andThen(({ createEntry, deleteEntry }) =>
      Effect.acquireRelease(createEntry(bucket, index), (entry, exit) =>
        Exit.isFailure(exit) ? deleteEntry(entry) : Effect.void
      )
    )
  )

export const make = Effect.scoped(
  Effect.Do.pipe(
    Effect.bind("bucket", () => createBucket),
    Effect.bind("index", () => createIndex),
    Effect.andThen(({ bucket, index }) => createEntry(bucket, index))
  )
)

// @filename: WorkspaceTest.ts
// ---cut---
import { Effect, Context, Layer, Console } from "effect"
import * as Services from "./Services"
import * as Workspace from "./Workspace"

// `FailureCaseLiterals` 型は、サービスをテストする際に異なるエラーシナリオを提供することを可能にします。
// 例えば、値として "S3" を提供することで、S3 サービスに特有のエラーシナリオをシミュレートできます。
// これにより、プログラムがエラーを正しく処理し、さまざまな状況で期待通りに動作することを確認できます。
// 同様に、"ElasticSearch" や "Database" などの値を提供することで、それらのサービスのエラーシナリオをシミュレートできます。
// エラーが発生しない場合をテストしたい場合は、`undefined` を提供できます。
// このパラメータを使用することで、サービスを徹底的にテストし、異なるエラー条件下での動作を検証できます。
type FailureCaseLiterals = "S3" | "ElasticSearch" | "Database" | undefined

class FailureCase extends Context.Tag("FailureCase")<
  FailureCase,
  FailureCaseLiterals
>() {}

// S3サービスのテストレイヤーを作成する

const S3Test = Layer.effect(
  Services.S3,
  Effect.andThen(FailureCase, (failureCase) => ({
    createBucket: Console.log("[S3] creating bucket").pipe(
      Effect.andThen(
        failureCase === "S3"
          ? Effect.fail(new Services.S3Error())
          : Effect.succeed({ name: "<bucket.name>" })
      )
    ),
    deleteBucket: (bucket) => Console.log(`[S3] delete bucket ${bucket.name}`)
  }))
)

// ElasticSearchサービスのテストレイヤーを作成する

const ElasticSearchTest = Layer.effect(
  Services.ElasticSearch,
  Effect.andThen(FailureCase, (failureCase) => ({
    createIndex: Console.log("[ElasticSearch] creating index").pipe(
      Effect.andThen(
        failureCase === "ElasticSearch"
          ? Effect.fail(new Services.ElasticSearchError())
          : Effect.succeed({ id: "<index.id>" })
      )
    ),
    deleteIndex: (index) =>
      Console.log(`[ElasticSearch] delete index ${index.id}`)
  }))
)

// Databaseサービスのテストレイヤーを作成する

const DatabaseTest = Layer.effect(
  Services.Database,
  Effect.andThen(FailureCase, (failureCase) => ({
    createEntry: (bucket, index) =>
      Console.log(
        `[Database] creating entry for bucket ${bucket.name} and index ${index.id}`
      ).pipe(
        Effect.andThen(
          failureCase === "Database"
            ? Effect.fail(new Services.DatabaseError())
            : Effect.succeed({ id: "<entry.id>" })
        )
      ),
    deleteEntry: (entry) => Console.log(`[Database] delete entry ${entry.id}`)
  }))
)

// S3、ElasticSearch、およびDatabaseサービスのすべてのテストレイヤーを単一のレイヤーにマージする

const layer = Layer.mergeAll(S3Test, ElasticSearchTest, DatabaseTest)

// Workspaceコードをテストするための実行可能なエフェクトを作成する
// エフェクトはテストレイヤーと未定義の値（失敗ケースなし）のFailureCaseサービスを提供される

const runnable = Workspace.make.pipe(
  Effect.provide(layer),
  Effect.provideService(FailureCase, undefined)
)

Effect.runPromise(Effect.either(runnable)).then(console.log)
```

</Tab>
</Tabs>

`FailureCase` が `undefined` に設定されているシナリオ（ハッピーパス）のテスト結果を見てみましょう：

```bash
[S3] creating bucket
[ElasticSearch] creating index
[Database] creating entry for bucket <bucket.name> and index <index.id>
{
  _id: "Either",
  _tag: "Right",
  right: {
    id: "<entry.id>"
  }
}
```

この場合、すべての操作が成功し、`right({ id: '<entry.id>' })` という成功した結果が表示されます。

次に、`Database`での失敗をシミュレートしてみましょう：

```ts
const runnable = Workspace.make.pipe(
  Effect.provide(layer),
  Effect.provideService(FailureCase, "Database")
)
```

The console output will be:

```bash
[S3] creating bucket
[ElasticSearch] creating index
[Database] creating entry for bucket <bucket.name> and index <index.id>
[ElasticSearch] delete index <index.id>
[S3] delete bucket <bucket.name>
{
  _id: "Either",
  _tag: "Left",
  left: {
    _tag: "DatabaseError"
  }
}
```

`Database`エラーが発生すると、まず`ElasticSearch`インデックスが削除され、その後関連する`S3`バケットが削除される完全なロールバックが行われることがわかります。結果は`left(new DatabaseError())`という失敗になります。

次に、インデックス作成を失敗させてみましょう：

```ts
const runnable = Workspace.make.pipe(
  Effect.provide(layer),
  Effect.provideService(FailureCase, "ElasticSearch")
)
```

この場合、コンソール出力は次のようになります：

```bash
[S3] creating bucket
[ElasticSearch] creating index
[S3] delete bucket <bucket.name>
{
  _id: "Either",
  _tag: "Left",
  left: {
    _tag: "ElasticSearchError"
  }
}
```

予想通り、`ElasticSearch`のインデックス作成が失敗すると、`S3`バケットが削除されるロールバックが行われます。結果は`left(new ElasticSearchError())`という失敗になります。
